Stored recipient list

kevinansfield opened the pull request:
We want to store a list of recipients for each bulk email so that we have a consistent set of data that background processing/sending jobs can work from without worrying about moving large data sets around or member data changing mid-send.

- added schema+migration for email_recipients table 
    - `email_recipients` table acts as a join table from email<->member
    - `member_id` does not have a foreign key constraint because members can be deleted but does have an index so that we can efficiently query which emails a member has received
    - stores static copies of the member info present at the time of sending an email for consistency in background jobs and auditing/historical data
- store the recipient list before sending an email
    - chunk already-fetched members list into batches and insert records into the `email_recipients` table via knex
    - chunked into batches of 1000 to match the number of emails that Mailgun accepts in a single API request but this may not be the absolute fastest batch size for recipient insertion:
        | Batch size | Batch time | Total time |
        | ---------- | ---------- | ---------- |
        |        500 |       20ms |     4142ms |
        |       1000 |       50ms |     4651ms |
        |       5000 |      170ms |     3540ms |
        |      10000 |      370ms |     3684ms |
- Fetch member data in mega without bookshelf  …
    - bookshelf can add around 3x overhead when fetching the members list for an email
    - we don't need full members at this point, only having the data is fine
    - if we need full models later on we can push the model hydration into background jobs where recipient batches are fetched ready for an email to be sent

kevinansfield committed:
Added schema+migration for email_recipients table

no issue

We want to store a list of recipients for each bulk email so that we have a consistent set of data that background processing/sending jobs can work from without worrying about moving large data sets around or member data changing mid-send.

- `email_recipients` table acts as a join table from email<->member
- `member_id` does not have a foreign key constraint because members can be deleted but does have an index so that we can efficiently query which emails a member has received
- stores static copies of the member info present at the time of sending an email for consistency in background jobs and auditing/historical data

kevinansfield committed:
Store recipient list before sending email

- chunk already-fetched members list into batches and insert records into the `email_recipients` table via knex
- chunked into batches of 1000 to match the number of emails that Mailgun accepts in a single API request but this may not be the absolute fastest batch size for recipient insertion:

| Batch size | Batch time | Total time |
| ---------- | ---------- | ---------- |
|        500 |       20ms |     4142ms |
|       1000 |       50ms |     4651ms |
|       5000 |      170ms |     3540ms |
|      10000 |      370ms |     3684ms |

kevinansfield committed:
Fetch member data in mega without bookshelf

- bookshelf can add around 3x overhead when fetching the members list for an email
- we don't need full members at this point, only having the data is fine
- if we need full models later on we can push the model hydration into background jobs where recipient batches are fetched ready for an email to be sent

kevinansfield commented on core/server/data/schema/schema.js:
```diff
@@ -459,5 +459,16 @@ module.exports = {
+        member_id: {type: 'string', maxlength: 24, nullable: false, index: true},
+        batch: {type: 'integer', nullable: false, unsigned: true, index: true},
+        processed_at: {type: 'dateTime', nullable: true},
+        // TODO: should these be prefixed with `member_` to denote their origin?
+        uuid: {type: 'string', maxlength: 36, nullable: false, validations: {isUUID: true}},
+        email: {type: 'string', maxlength: 191, nullable: false, validations: {isEmail: true}},
+        name: {type: 'string', maxlength: 191, nullable: true}
```
@matthanley the `member_` prefix wasn't in the original db schema in https://github.com/TryGhost/Team/issues/353#issue-668959800 but I think it would be useful, especially for the `uuid` field because that could be quite confusing if you expect it to be a uuid but it's not unique across `email_recipients` rows. Do you think it's a change worth making?

kevinansfield commented on core/server/data/schema/schema.js:
```diff
@@ -459,5 +459,16 @@ module.exports = {
+    },
+    email_recipients: {
+        id: {type: 'string', maxlength: 24, nullable: false, primary: true},
+        email_id: {type: 'string', maxlength: 24, nullable: false, references: 'emails.id'},
+        member_id: {type: 'string', maxlength: 24, nullable: false, index: true},
+        batch: {type: 'integer', nullable: false, unsigned: true, index: true},
```
The indexes here could do with some further performance tests. I added an index to the `batch` column because the background jobs will be fetching with `where email_id="x" and batch=n` but it would be interesting to know if a composite index works better in this scenario. Although I'm not sure how we'd go about adding composite indexes with our schema tools 🤔

ErisDS commented:
Just dropping a note here that I've added myself as a reviewer as I'm looking at this :)

kevinansfield committed:
Added `email_batches` table

- replaced a `batch` number in the `email_recipients` table with a separate `email_batches` table and associated reference in the `email_recipients` column
- allows for storage of batch status, mailgun email id, and better per-batch error data

kevinansfield committed:
Added EmailBatch/EmailRecipient models

- added initial model definitions with defaults and relationships
- added missing `post` relationship function to email model

kevinansfield committed:
Stored batch records and attach recipient records when generating email list

- create a batch record before inserting recipient rows with the batch records id rather than storing a plain batch number

ErisDS commented on core/server/data/schema/schema.js:
```diff
@@ -459,5 +459,31 @@ module.exports = {
+        batch_id: {type: 'string', maxlength: 24, nullable: false, references: 'email_batches.id'},
+        processed_at: {type: 'dateTime', nullable: true},
+        delivered_at: {type: 'dateTime', nullable: true},
+        failed_at: {type: 'dateTime', nullable: true},
```
Thinking this through, failed is going to come with a reason, and I'm not sure that storing error_data on a batch is right either if it's still a big JSON blob? 

I think we have 2 options here, either spend a bit more time thinking about how this will actually get used in order to get the structure right first time, or pull these columns until we're sure how we would use them? 

Having to add columns later is easy. Having to modify columns later is a PITA.

kevinansfield replied:
> failed is going to come with a reason
Yeah, I thought that might be in whatever "activities" table we had later so that we have a history of attempts and associated errors, unless we plan to update this record based on events coming from Mailgun and only keep the latest error state until it's a success?

> I'm not sure that storing error_data on a batch is right either if it's still a big JSON blob
If it's just for auditing a JSON blob could be useful when there's no access to the logs. Seeing as this is an API error response that we want to store perhaps we should follow our own error pattern and have a message/details/context that the bulk email provider can populate by transforming the provider error?

ErisDS replied:
> I thought that might be in whatever "activities" table we had later so that we have a history of attempts and associated errors, unless we plan to update this record based on events coming from Mailgun and only keep the latest error state until it's a success?

This is IMO confusing permanent and temporary failures which are different. AFAIK there can be only one permanent fail per email.

> If it's just for auditing a JSON blob could be useful when there's no access to the logs. Otherwise, seeing as this is an API error response that we want to store perhaps we should follow our own error pattern and have a message/details/context that the bulk email provider can populate by transforming the provider error?

I think what's missing here is clarity on what the Mailgun API _can_ return?

kevinansfield replied:
> This is IMO confusing permanent and temporary failures which are different. AFAIK there can be only one permanent fail per email.

I think there can only be one permanent failure unless the email is manually retried from the Mailgun UI? I was thinking more about exposing that delivery attempts have been made+failed/are continuing so that member-site support staff know what's happening. Was the intention for the email_recipient record to show failed only for a permanent failure?

> I think what's missing here is clarity on what the Mailgun API can return?

Agreed. A quick scan of their docs doesn't provide anything more helpful than a description of HTTP status codes 😞 I guess we'd also need to consider errors occurring before the Mailgun API endpoint is reached such as network issues, firewalls, gateway timeouts etc?

ErisDS replied:
All of this is a much much bigger discussion - so I suggest going with option 1 and pulling the fields we're not sure about yet.
